lib/index.js

```diff
@@ -2,3 +2,163 @@
+    try {
+      const initialAccessToken = await _authenticate({
+        req, authentication, logger});
+
+      const {body: registrationBody} = req;
+
+      const registration = _parseRegistration({registrationBody, defaults});
+
+      const result = registration.validateNew();
+      if(!result.valid) {
+        throw result.error;
+      }
+
+      // client can be optionally overriden by the register() callback
+      await _ensureClientId({registration, allowClientProvidedId});
+      await _generateClientSecret({registration, defaults});
+
+      // registration.registration_access_token = '...';
+      // registration.registration_client_uri = (new URL('/register/' +
+      //   encodeURIComponent(registration.client_id), baseUrl)).toString();
+
+      // Epoch time
+      registration.client_id_issued_at = Math.floor(Date.now() / 1000);
+
+      let client;
+      try {
+        client = await register({
+          registration, credentials: {initialAccessToken}
+        });
+      } catch(e) {
+        if(logger) {
+          logger.error(e);
+        }
+        throw new InvalidRequest({
+          description: 'Error registering client.'
+        });
+      }
+
+      _respond({client, res});
+    } catch(error) {
+      if(logger) {
+        logger.error(error);
+      }
+      error.respond(res);
+    }
+  };
+}
+
+function _respond({client, res}) {
+  res.set({
+    'Cache-Control': 'no-store',
+    Pragma: 'no-cache'
+  });
+
+  res.status(201).json(client);
+}
+
+function _parseRegistration({registrationBody, defaults}) {
+  if(!registrationBody) {
+    throw new InvalidRequest({
+      error: 'invalid_request',
+      error_description: 'Missing registration request body.'
+    });
+  }
+
+  const hit = FORBIDDEN_FIELDS.find(prop =>
+    registrationBody[prop] !== undefined);
+  if(hit) {
+    throw new InvalidRequest({
+      description: `Registration MUST NOT include the ${hit} field.`
+    });
+  }
+
+  const registrationDefaults = {};
+
+  for(const [key, value] of Object.entries(defaults)) {
+    registrationDefaults[snakeCase(key)] = value;
+  }
+
+  return new ClientMetadata({...registrationDefaults, ...registrationBody});
+}
+
+async function _authenticate({req, res, authentication}) {
+  let token;
+
+  if(authentication.strategy === 'bearer') {
+    token = await _extractBearerToken({req, res});
+    try {
+      await authentication.validateInitialAccessToken({token});
+    } catch(e) {
+      throw new AccessDenied({
+        description: `Invalid authentication code: ${e.message}`
+      });
+    }
+
+  } else {
+    throw new InvalidRequest({
+      description: 'Only the Bearer auth scheme is currently supported.'});
+  }
+  return token;
+}
+
+async function _extractBearerToken({req}) {
+  const authorization = req.headers && req.headers.authorization;
+
+  if(!authorization) {
+    throw new AccessDenied({description: 'Authentication Code required.'});
+  }
+  const components = authorization.split(' ');
```

Comment:
can that parsing lib used elsewhere be used here as well?

Reply:
Yes, we can refactor to use that lib.

Reply:
Refactored, using `auth-header` now.